### DISCUSSION TITLE
Certbot automatic renewal

### DISCUSSION THREAD

**Discussion by inventor96** [NightscoutFoundation/xDrip#2518] [Q&A] [answered]
I was just reviewing the setup scripts that were created for the migration to Google Cloud Platform (which are awesome, btw), and noticed that I can't seem to find anything about certbot certificate renewals. Certificates obtained using certbot last for 3 months at a time (see my example from my NS site created today).

![image](https://user-images.githubusercontent.com/7132744/205424926-5df34ce1-4a3b-4c7a-9bc6-464adc552a24.png)

So renewals need to happen more frequently than most users will want to do manually. A simple `certbot renew` command in the crontab for once or twice a month will do (I personally do it on days 1 and 15 monthly wherever I use certbot).

I've tried searching the discussions to see if this was already addressed, but I couldn't find anything on it. If I missed it, please point me in the right direction!

**Comment by Navid200** [comment]
@tzachi-dar @jamorham

**Comment by Navid200** [comment]
@inventor96 certbot is renewed automatically, isn't it?

You can type the following to verify:

sudo certbot renew --dry-run

**Comment by tzachi-dar** [comment]
certbot creates the file sudo cat /etc/cron.d/certbot that should solve the issue.
But, we never installed cron, so it might not be working.
The best way to understand if this is an issue, is to look at a machine that is running for more than a month and see if the certificate was renewed there.
Navid, do you have such a machine? Can you check the certificate using chrome?

**Reply by Navid200** [reply]
My machine stopped yesterday because I had not upgraded my Google Project yet:  https://navid200.github.io/xDrip/docs/Nightscout/FreeTrialEnd.html

In order to bring it back up, I had to run the FreeDNS Setup utility on it.
So, certbot has been executed on it just yesterday.

**Comment by tzachi-dar** [comment]
All the machines that I'm aware of are still too new for this. It says something like:
Cert not yet due for renewal
I'll ask someone with a machine that is installed for more than 30 days tomorrow and let me know how that works.

**Comment by inventor96** [comment]
My knowledge could be old, but my understanding was that certbot did not come with automatic renewal. If someone wanted automatic renewal, an entry in the crontab to call `certbot renew` in some fashion was required.

@tzachi-dar crond comes by default on the Ubuntu version recommended in the setup instructions, so that's not a worry. Adding it to the installation steps would be a good thing for those non-standard situations.

I don't think certs will renew after 30 days of deployment... I believe it's the other way around; it won't renew until it's less than 30 days until the expiration. See https://github.com/certbot/certbot/blob/f6d532a15bdaf55e2fd6ba630075aaa22c5d8cfc/certbot/certbot/_internal/constants.py#L153

**Comment by inventor96** [comment]
@tzachi-dar I didn't think until just now to look at the file you referenced. This is the contents of that file on my system:

```
$ cat /etc/cron.d/certbot
# /etc/cron.d/certbot: crontab entries for the certbot package
#
# Upstream recommends attempting renewal twice a day
#
# Eventually, this will be an opportunity to validate certificates
# haven't been revoked, etc.  Renewal will only occur if expiration
# is within 30 days.
#
# Important Note!  This cronjob will NOT be executed if you are
# running systemd as your init system.  If you are running systemd,
# the cronjob.timer function takes precedence over this cronjob.  For
# more details, see the systemd.timer manpage, or use systemctl show# certbot.timer.
SHELL=/bin/sh
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin

0 */12 * * * root test -x /usr/bin/certbot -a \! -d /run/systemd/system && perl -e 'sleep int(rand(43200))' && certbot -q renew
```

Ubuntu 20 does use systemd, so it looks like the timer needs enabled instead of using crond like we've discussed so far.

**Comment by tzachi-dar** [comment]
Well, the oldest site that I know of is one month old. The certificate has not updated, but that might still happen until tomorrow. I'll wait one more day and see if this is an issue or not.

**Reply by inventor96** [reply]
We wouldn't know if it's a problem until the site (more specifically the certificate) is at least 2 months old. Certbot doesn't renew certs until it's less than 30 days until the expiration.

**Comment by tzachi-dar** [comment, marked as answer]
Thanks @inventor96 I thought they do it after a month, but you are right.
In any case, I have looked at one of the systems, and they seem to have logs saying that renewal process is running. So I'm 99.99% percent sure this will be renewed when the time comes.

 sudo cat /var/log/letsencrypt/letsencrypt.log
2022-12-04 13:46:22,906:DEBUG:certbot.main:certbot version: 0.40.0
2022-12-04 13:46:22,907:DEBUG:certbot.main:Arguments: ['-q']
2022-12-04 13:46:22,907:DEBUG:certbot.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#manual,PluginEntryPoint#nginx,PluginEntryPoint#null,PluginEntryPoint#standalone,PluginEntryPoint#webroot)
2022-12-04 13:46:23,006:DEBUG:certbot.log:Root logging level set at 30
2022-12-04 13:46:23,007:INFO:certbot.log:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2022-12-04 13:46:23,049:DEBUG:certbot.plugins.selection:Requested authenticator <certbot.cli._Default object at 0x7ff59843dca0> and installer <certbot.cli._Default object at 0x7ff59843dca0>
2022-12-04 13:46:23,059:INFO:certbot.renewal:Cert not yet due for renewal
2022-12-04 13:46:23,059:DEBUG:certbot.plugins.selection:Requested authenticator nginx and installer nginx
2022-12-04 13:46:23,060:DEBUG:certbot.plugins.selection:Selecting plugin: * nginx
Description: Nginx Web Server plugin
Interfaces: IAuthenticator, IInstaller, IPlugin
Entry point: nginx = certbot_nginx.configurator:NginxConfigurator
Initialized: <certbot_nginx.configurator.NginxConfigurator object at 0x7ff5983edc10>
2022-12-04 13:46:23,128:INFO:certbot.renewal:Cert not yet due for renewal
2022-12-04 13:46:23,129:DEBUG:certbot.plugins.selection:Requested authenticator nginx and installer nginx
2022-12-04 13:46:23,129:DEBUG:certbot.plugins.selection:Selecting plugin: * nginx
Description: Nginx Web Server plugin
Interfaces: IAuthenticator, IInstaller, IPlugin
Entry point: nginx = certbot_nginx.configurator:NginxConfigurator
Initialized: <certbot_nginx.configurator.NginxConfigurator object at 0x7ff5983ed550>
2022-12-04 13:46:23,130:DEBUG:certbot.renewal:no renewal failures

**Reply by inventor96** [reply]
Ahh, good thinking. That's interesting it's working like that, since it's contrary to the notes in the certbot crond file. Oh well. I think this answers the question, so we'll just leave this be unless we start hearing issues about expired certificates!
